third_party/kiwi/kiwi/parser_pgoptions.c

```diff
@@ -0,0 +1,100 @@
+
+/*
+ * kiwi.
+ *
+ * postgreSQL protocol interaction library.
+ */
+
+#include <getopt.h>
+#include <wordexp.h>
+#include <assert.h>
+
+#include "kiwi.h"
+
+static void
+ParseLongOption(const char *string, char **name, char **value)
+{
+	size_t equal_pos;
+	char *cp;
+
+	assert(string);
+	assert(name);
+	assert(value);
+
+	equal_pos = strcspn(string, "=");
+
+	if (string[equal_pos] == '=') {
+		*name = malloc(equal_pos + 1);
+		strncpy(*name, string, equal_pos);
+
+		*value = strdup(&string[equal_pos + 1]);
+	} else {
+		/* no equal sign in string */
+		*name  = strdup(string);
+		*value = NULL;
+	}
+
+	for (cp = *name; *cp; cp++)
+		if (*cp == '-')
+			*cp = '_';
+}
+
+int
+kiwi_parse_pgoptions_and_update_vars(kiwi_vars_t *vars,
+                                     char *pgoptions,
+                                     int pgoptions_len)
+{
+	int errs = 0;
+	int flag;
+
+	int argc;
+	char **argv;
+	char *old_pgoptions = NULL;
+
+	if (pgoptions[pgoptions_len - 1] != '\0') {
+		char *old_pgoptions = pgoptions;
+		pgoptions           = malloc(pgoptions_len + 1);
+		memcpy(pgoptions, old_pgoptions, pgoptions_len);
+		pgoptions[pgoptions_len] = '\0';
+	}
+
+	wordexp_t res;
+	wordexp(pgoptions, &res, 0);
+
+	argc = res.we_wordc + 1;
+	argv = malloc(sizeof(char *) * (argc + 1));
+	if (argv == 0) {
+		return -1;
+	}
+
+	argv[0] = "fake_word"; // needs for getopt function
+
+	memcpy(&argv[1], res.we_wordv, sizeof(char *) * argc);
```

Comment:
Do not forget to place that 0 byte at the end.

Reply:
`res.we_wordv` yet contains 0 at the end.